src/commonMain/kotlin/fr/acinq/lightning/wire/OfferTypes.kt

```diff
@@ -0,0 +1,1003 @@
+package fr.acinq.lightning.wire
+
+import fr.acinq.bitcoin.*
+import fr.acinq.bitcoin.io.ByteArrayInput
+import fr.acinq.bitcoin.io.ByteArrayOutput
+import fr.acinq.bitcoin.io.Input
+import fr.acinq.bitcoin.io.Output
+import fr.acinq.lightning.*
+import fr.acinq.lightning.Lightning.randomBytes32
+import fr.acinq.lightning.crypto.RouteBlinding
+import fr.acinq.lightning.utils.Either
+import fr.acinq.lightning.utils.Either.*
+
+/**
+ * Lightning Bolt 12 offers
+ * see https://github.com/lightning/bolts/blob/master/12-offer-encoding.md
+ */
+object OfferTypes {
+    /** Data provided to reach the issuer of an offer or invoice. */
+    sealed interface ContactInfo {
+        /** If the offer or invoice issuer doesn't want to hide their identity, they can directly share their public nodeId. */
+        data class RecipientNodeId(val nodeId: PublicKey) : ContactInfo
+
+        /** If the offer or invoice issuer wants to hide their identity, they instead provide blinded paths. */
+        data class BlindedPath(val route: RouteBlinding.BlindedRoute) : ContactInfo
+    }
+
+    fun writePath(path: ContactInfo.BlindedPath, out: Output) {
+        LightningCodecs.writeEncodedNodeId(path.route.introductionNodeId, out)
+        LightningCodecs.writeBytes(path.route.blindingKey.value, out)
+        LightningCodecs.writeByte(path.route.blindedNodes.size, out)
+        for (node in path.route.blindedNodes) {
+            LightningCodecs.writeBytes(node.blindedPublicKey.value, out)
+            LightningCodecs.writeU16(node.encryptedPayload.size(), out)
+            LightningCodecs.writeBytes(node.encryptedPayload, out)
+        }
+    }
+
+    fun readPath(input: Input): ContactInfo.BlindedPath {
+        val introductionNodeId = LightningCodecs.encodedNodeId(input)
+        val blindingKey = PublicKey(LightningCodecs.bytes(input, 33))
+        val blindedNodes = ArrayList<RouteBlinding.BlindedNode>()
+        val numBlindedNodes = LightningCodecs.byte(input)
+        for (i in 1 .. numBlindedNodes) {
+            val blindedKey = PublicKey(LightningCodecs.bytes(input, 33))
+            val payload = ByteVector(LightningCodecs.bytes(input, LightningCodecs.u16(input)))
+            blindedNodes.add(RouteBlinding.BlindedNode(blindedKey, payload))
+        }
+        return ContactInfo.BlindedPath(RouteBlinding.BlindedRoute(introductionNodeId, blindingKey, blindedNodes))
+    }
+
+    sealed interface Bolt12Tlv : Tlv
+
+    sealed interface InvoiceTlv : Bolt12Tlv
+
+    sealed interface InvoiceRequestTlv : InvoiceTlv
+
+    sealed interface OfferTlv : InvoiceRequestTlv
+
+    sealed interface InvoiceErrorTlv : Bolt12Tlv
```

Comment:
Same here, those should be `sealed class` (I believe this should work the same, you won't need to nest the types inside the `sealed class`).